src/lib/bootloader/grub2pwd.rb

```diff
@@ -1,63 +1,131 @@
-  PWD_ENCRYPTION_FILE = "/etc/grub.d/42_password"
+Yast.import "Stage"
 
-  def used?
-    Yast.import "FileUtils"
+module Bootloader
+  # class is responsible for detection, encryption and writing of grub2 password protection
+  class GRUB2Pwd
+    attr_accessor :used, :unrestricted
+    alias_method :used?, :used
+    alias_method :unrestricted?, :unrestricted
 
-    Yast::FileUtils.Exists PWD_ENCRYPTION_FILE
-  end
+    def initialize
+      # TODO: offline upgrade should somehow read it when we need grub2 -> grub2 modifications
+      if Yast::Stage.initial
+        propose
+      else
+        read
+      end
+    end
 
-  def enable(password)
-    enc_passwd = encrypt(password)
-
-    file_content = "#! /bin/sh\n" \
-      "exec tail -n +3 $0\n" \
-      "# File created by YaST and next password change in YaST will overwrite it\n" \
-      "set superusers=\"root\"\n" \
-      "password_pbkdf2 root #{enc_passwd}\n" \
-      "export superusers"
-
-    Yast::SCR.Write(
-      Yast::Path.new(".target.string"),
-      [PWD_ENCRYPTION_FILE, 0700],
-      file_content
-    )
-  end
+    def write
+      if used?
+        enable
+      else
+        disable
+      end
+    end
 
-  def disable
-    return unless used?
+    def password=(value)
+      @encrypted_password = encrypt(value)
+    end
 
-    Yast::SCR.Execute(YAST_BASH_PATH, "rm '#{PWD_ENCRYPTION_FILE}'")
-  end
+    def password?
+      !@encrypted_password.nil?
+    end
+
+  private
+
+    YAST_BASH_PATH = Yast::Path.new(".target.bash_output")
+    PWD_ENCRYPTION_FILE = "/etc/grub.d/42_password"
+
+    def propose
+      @used = false
+      @unrestricted = false # TODO: ensure it in FATE
+      @encrypted_password = nil # not set by default
+    end
+
+    def read
+      if !read_used
+        propose
+        return
+      end
+
+      @used = true
+      content = Yast::SCR.Read(
+        Yast::Path.new(".target.string"),
+        PWD_ENCRYPTION_FILE
+      )
 
-private
+      unrestricted_lines = content.lines.grep(/unrestricted_menuentry_users/)
+      @unrestricted = !unrestricted_lines.empty?
 
-  def encrypt(password)
-    Yast.import "String"
+      pwd_line = content.lines.grep(/password_pbkdf2 root/).first
 
-    quoted_password = Yast::String.Quote(password)
-    result = Yast::WFM.Execute(YAST_BASH_PATH,
-      "echo '#{quoted_password}\n#{quoted_password}\n' | LANG=C grub2-mkpasswd-pbkdf2"
-    )
+      if !pwd_line
+        raise "Cannot find encrypted password, YaST2 password generator in /etc/grub.d is probably modified."
+      end
 
-    if result["exit"] != 0
-      raise "Failed to create encrypted password for grub2. Command output: #{result["stderr"]}"
+      @encrypted_password = pwd_line[/password_pbkdf2 root (\S+)/, 1]
     end
 
-    pwd_line = result["stdout"].split("\n").grep(/password is/).first
-    if !pwd_line
-      raise "INTERNAL ERROR: output do not contain encrypted password. Output: #{result["stdout"]}"
+    def read_used
```

Comment:
So the intent is "read the value for the attribute `used` from permanent storage but don't assign it". 

The yast convention is that Read should assign the attributes (and the preceding methods conforms to that).

Maybe we can set a convention that Read without assigning is called Load. -> `load_used`